src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,675 @@
+ *     }
+ *
+ * }
+ *}
+ *
+ * <h2 id="stable-collections">Stable Collections</h2>
+ * Stable values can also be used as backing storage for
+ * {@linkplain Collection##unmodifiable unmodifiable collections}. A <em>stable list</em>
+ * is an unmodifiable list, backed by an array of stable values. The stable list elements
+ * are computed when they are first accessed, using a provided {@linkplain IntFunction}:
+ *
+ * {@snippet lang = java:
+ * public final class SqrtUtil {
+ *
+ *     private SqrtUtil(){}
+ *
+ *     private static final List<Double> SQRT =
+ *             // @link substring="list" target="#list(int,IntFunction)" :
+ *             StableValue.list(10, StrictMath::sqrt);
+ *
+ *     public static double sqrt9() {
+ *         return SQRT.get(9); // May eventually constant fold to 3.0 at runtime
+ *     }
+ *
+ * }
+ *}
+ * <p>
+ * Similarly, a <em>stable map</em> is an unmodifiable map whose keys are known at
+ * construction. The stable map values are computed when they are first accessed,
+ * using a provided {@linkplain Function}:
+ *
+ * {@snippet lang = java:
+ * public final class SqrtUtil {
+ *
+ *     private SqrtUtil(){}
+ *
+ *     private static final Map<Integer, Double> SQRT =
+ *             // @link substring="map" target="#map(Set,Function)" :
+ *             StableValue.map(Set.of(1, 2, 4, 8, 16, 32), StrictMath::sqrt);
+ *
+ *     public static double sqrt16() {
+ *         return SQRT.get(16); // May eventually constant fold to 4.0 at runtime
+ *     }
+ *
+ * }
+ *}
+ *
+ * <h2 id="composition">Composing stable values</h2>
+ * A stable value can depend on other stable values, thereby creating a dependency graph
+ * that can be lazily computed but where access to individual elements still can be
+ * constant-folded. In the following example, a single {@code Foo} and a {@code Bar}
+ * instance (that is dependent on the {@code Foo} instance) are lazily created, both of
+ * which are held by stable values:
+ * {@snippet lang = java:
+ * public final class DependencyUtil {
+ *
+ *     private DependencyUtil(){}
+ *
+ *     public static class Foo {
+ *          // ...
+ *      }
+ *
+ *     public static class Bar {
+ *         public Bar(Foo foo) {
+ *              // ...
+ *         }
+ *     }
+ *
+ *     private static final Supplier<Foo> FOO = StableValue.supplier(Foo::new);
+ *     private static final Supplier<Bar> BAR = StableValue.supplier(() -> new Bar(FOO.get()));
+ *
+ *     public static Foo foo() {
+ *         return FOO.get();
+ *     }
+ *
+ *     public static Bar bar() {
+ *         return BAR.get();
+ *     }
+ *
+ * }
+ *}
+ * Calling {@code bar()} will create the {@code Bar} singleton if it is not already
+ * created. Upon such a creation, the dependent {@code Foo} will first be created if
+ * the {@code Foo} does not already exist.
+ * <p>
+ * Here is another example where a more complex dependency graph is created in which
```

Comment:
I wonder if just leaving the fibonacci example would be enough here -- as that has a nice dependency graph